{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2fed000",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "55bab975",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import anndata\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import post_preprocess as ppp\n",
    "import re\n",
    "    \n",
    "\"\"\"\n",
    "This loads and preprocesses the Lincs L1000 data\n",
    "\n",
    "https://lincsproject.org/LINCS/tools/workflows/find-the-best-place-to-obtain-the-lincs-l1000-data\n",
    "https://www.cell.com/cell/pdf/S0092-8674(17)31309-0.pdf\n",
    "\n",
    "I downloaded the anndata from chemCPA, from https://dl.fbaipublicfiles.com/dlp/cpa_binaries.tar, \n",
    "\"\"\"\n",
    "\n",
    "def remove_non_alphanumeric(input_string):\n",
    "    return re.sub(r\"[^a-zA-Z0-9]\", \"\", input_string)\n",
    "\n",
    "load_path='/home/manu/chemCPA/chemCPA/anndatas'\n",
    "adata=sc.read(f'{load_path}/lincs_full.h5ad')\n",
    "\n",
    "# It's not single cell expression data, so I will not normalize+log1p it\n",
    "\n",
    "#sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "#sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "71aa5316",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>treatment</th>\n",
       "      <th>treatment_dose_uM</th>\n",
       "      <th>treatment_time_h</th>\n",
       "      <th>treatment_type</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>REP.A001_A375_24H_X1_B22:A03-2</th>\n",
       "      <td>A375</td>\n",
       "      <td>control</td>\n",
       "      <td>0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REP.A001_A375_24H_X1_B22:A04-2</th>\n",
       "      <td>A375</td>\n",
       "      <td>control</td>\n",
       "      <td>0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REP.A001_A375_24H_X1_B22:A05-2</th>\n",
       "      <td>A375</td>\n",
       "      <td>control</td>\n",
       "      <td>0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REP.A001_A375_24H_X1_B22:A06-2</th>\n",
       "      <td>A375</td>\n",
       "      <td>control</td>\n",
       "      <td>0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REP.A001_A375_24H_X1_B22:A07-2</th>\n",
       "      <td>A375</td>\n",
       "      <td>BRDK25114078</td>\n",
       "      <td>10.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCLB003_PC3_24H_X3_B13:P20-1</th>\n",
       "      <td>PC3</td>\n",
       "      <td>BRDA75409952</td>\n",
       "      <td>3.32999992371</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCLB003_PC3_24H_X3_B13:P21-1</th>\n",
       "      <td>PC3</td>\n",
       "      <td>BRDA75409952</td>\n",
       "      <td>1.1100000143100002</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCLB003_PC3_24H_X3_B13:P22-1</th>\n",
       "      <td>PC3</td>\n",
       "      <td>BRDA75409952</td>\n",
       "      <td>0.3700000047680001</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCLB003_PC3_24H_X3_B13:P23-1</th>\n",
       "      <td>PC3</td>\n",
       "      <td>BRDA75409952</td>\n",
       "      <td>0.119999997318</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCLB003_PC3_24H_X3_B13:P24-1</th>\n",
       "      <td>PC3</td>\n",
       "      <td>BRDA75409952</td>\n",
       "      <td>0.0399999991059</td>\n",
       "      <td>24.0</td>\n",
       "      <td>drug_perturbation</td>\n",
       "      <td>L1000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1034271 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cell_line     treatment   treatment_dose_uM  \\\n",
       "REP.A001_A375_24H_X1_B22:A03-2      A375       control                   0   \n",
       "REP.A001_A375_24H_X1_B22:A04-2      A375       control                   0   \n",
       "REP.A001_A375_24H_X1_B22:A05-2      A375       control                   0   \n",
       "REP.A001_A375_24H_X1_B22:A06-2      A375       control                   0   \n",
       "REP.A001_A375_24H_X1_B22:A07-2      A375  BRDK25114078                10.0   \n",
       "...                                  ...           ...                 ...   \n",
       "PCLB003_PC3_24H_X3_B13:P20-1         PC3  BRDA75409952       3.32999992371   \n",
       "PCLB003_PC3_24H_X3_B13:P21-1         PC3  BRDA75409952  1.1100000143100002   \n",
       "PCLB003_PC3_24H_X3_B13:P22-1         PC3  BRDA75409952  0.3700000047680001   \n",
       "PCLB003_PC3_24H_X3_B13:P23-1         PC3  BRDA75409952      0.119999997318   \n",
       "PCLB003_PC3_24H_X3_B13:P24-1         PC3  BRDA75409952     0.0399999991059   \n",
       "\n",
       "                                treatment_time_h     treatment_type dataset  \n",
       "REP.A001_A375_24H_X1_B22:A03-2              24.0  drug_perturbation   L1000  \n",
       "REP.A001_A375_24H_X1_B22:A04-2              24.0  drug_perturbation   L1000  \n",
       "REP.A001_A375_24H_X1_B22:A05-2              24.0  drug_perturbation   L1000  \n",
       "REP.A001_A375_24H_X1_B22:A06-2              24.0  drug_perturbation   L1000  \n",
       "REP.A001_A375_24H_X1_B22:A07-2              24.0  drug_perturbation   L1000  \n",
       "...                                          ...                ...     ...  \n",
       "PCLB003_PC3_24H_X3_B13:P20-1                24.0  drug_perturbation   L1000  \n",
       "PCLB003_PC3_24H_X3_B13:P21-1                24.0  drug_perturbation   L1000  \n",
       "PCLB003_PC3_24H_X3_B13:P22-1                24.0  drug_perturbation   L1000  \n",
       "PCLB003_PC3_24H_X3_B13:P23-1                24.0  drug_perturbation   L1000  \n",
       "PCLB003_PC3_24H_X3_B13:P24-1                24.0  drug_perturbation   L1000  \n",
       "\n",
       "[1034271 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cell_line']=adata.obs['cell_id']\n",
    "adata.obs['treatment']=adata.obs['pert_id']\n",
    "adata.obs['treatment']=[a if a!='DMSO' else 'control' for a in adata.obs['treatment']]\n",
    "adata.obs[\"treatment\"] = adata.obs[\"treatment\"].apply(remove_non_alphanumeric)\n",
    "\n",
    "adata.obs['treatment_dose_uM']=adata.obs['pert_dose'].astype('str')\n",
    "adata.obs['treatment_dose_uM']=[a if a!='-666.0' else '0' for a in adata.obs['treatment_dose_uM']]\n",
    "adata.obs['treatment_time_h']=adata.obs['pert_time']\n",
    "adata.obs['treatment_type']='drug_perturbation'\n",
    "adata.obs['dataset']='L1000'\n",
    "\n",
    "adata.obs=adata.obs[['cell_line', 'treatment', 'treatment_dose_uM', 'treatment_time_h', 'treatment_type', 'dataset']]\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "dda3ea3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map gene symbols to Ensemble IDs\n",
    "import os\n",
    "current_directory = os.getcwd()\n",
    "data_path='/'+os.path.join(*current_directory.split('/')[:-1])+'/non_anndata_data'\n",
    "\n",
    "Gene_Dict=np.load(f'{data_path}/Gene_Dict.npy', allow_pickle=True).item()\n",
    "adata.var['Gene_symbol_reduced']=[remove_non_alphanumeric(a) for a in adata.var.index]\n",
    "adata.var['Ensemble_ID']=[Gene_Dict[a] for a in adata.var['Gene_symbol_reduced']]\n",
    "adata.var.index=adata.var['Ensemble_ID']\n",
    "\n",
    "for c in adata.var.columns:\n",
    "    del adata.var[c]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6efdd09",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22b03879",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b821ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "448f16e5",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/manu/chemCPA/chemCPA/Preprocessing/post_preprocess.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  adata.obs['treatment']=[Specific_dict[a] for a in adata.obs['treatment']]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Treatments kept: 15860\n",
      "Treatments removed: 5444\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/manu/.local/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A375\n",
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n",
      "HA1E\n",
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HELA\n",
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/manu/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1222: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'columns'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m adata\u001b[38;5;241m=\u001b[39mppp\u001b[38;5;241m.\u001b[39mmap_and_filter_based_on_embedding(adata)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# This calls DEGs and filters out treatments with to little cells\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m adata\u001b[38;5;241m=\u001b[39m\u001b[43mppp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpostprocess_anndata\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_top_genes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2000\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/chemCPA/chemCPA/Preprocessing/post_preprocess.py:100\u001b[0m, in \u001b[0;36mpostprocess_anndata\u001b[0;34m(adata, n_top_genes)\u001b[0m\n\u001b[1;32m     98\u001b[0m         cell_line_kept\u001b[38;5;241m.\u001b[39mappend(cell_line)\n\u001b[1;32m     99\u001b[0m adata\u001b[38;5;241m=\u001b[39madata[adata\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell_line\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(cell_line_kept)]\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[0;32m--> 100\u001b[0m \u001b[43mrank_genes_groups_by_cov\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroupby\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpert_category\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontrol_group\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcontrol\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcovariate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcell_line\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    103\u001b[0m \u001b[38;5;66;03m# Split into train/test/ood sets\u001b[39;00m\n\u001b[1;32m    104\u001b[0m adata\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msplit\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtrain\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "File \u001b[0;32m~/chemCPA/chemCPA/Preprocessing/post_preprocess.py:194\u001b[0m, in \u001b[0;36mrank_genes_groups_by_cov\u001b[0;34m(adata, groupby, control_group, covariate, pool_doses, n_genes, rankby_abs, key_added, return_dict)\u001b[0m\n\u001b[1;32m    191\u001b[0m adata_cov \u001b[38;5;241m=\u001b[39m adata[adata\u001b[38;5;241m.\u001b[39mobs[covariate] \u001b[38;5;241m==\u001b[39m cov_cat]\n\u001b[1;32m    193\u001b[0m \u001b[38;5;66;03m# compute DEGs\u001b[39;00m\n\u001b[0;32m--> 194\u001b[0m \u001b[43msc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank_genes_groups\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    195\u001b[0m \u001b[43m    \u001b[49m\u001b[43madata_cov\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    196\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroupby\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreference\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcontrol_group_cov\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrankby_abs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrankby_abs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    199\u001b[0m \u001b[43m    \u001b[49m\u001b[43mn_genes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_genes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    200\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    202\u001b[0m \u001b[38;5;66;03m# add entries to dictionary of gene sets\u001b[39;00m\n\u001b[1;32m    203\u001b[0m de_genes \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(adata_cov\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrank_genes_groups\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:622\u001b[0m, in \u001b[0;36mrank_genes_groups\u001b[0;34m(adata, groupby, use_raw, groups, reference, n_genes, rankby_abs, pts, key_added, copy, method, corr_method, tie_correct, layer, **kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m test_obj\u001b[38;5;241m.\u001b[39mpts_rest \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    618\u001b[0m     adata\u001b[38;5;241m.\u001b[39muns[key_added][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpts_rest\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m    619\u001b[0m         test_obj\u001b[38;5;241m.\u001b[39mpts_rest\u001b[38;5;241m.\u001b[39mT, index\u001b[38;5;241m=\u001b[39mtest_obj\u001b[38;5;241m.\u001b[39mvar_names, columns\u001b[38;5;241m=\u001b[39mgroups_names\n\u001b[1;32m    620\u001b[0m     )\n\u001b[0;32m--> 622\u001b[0m test_obj\u001b[38;5;241m.\u001b[39mstats\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m \u001b[43mtest_obj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstats\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241m.\u001b[39mswaplevel()\n\u001b[1;32m    624\u001b[0m dtypes \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    625\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mO\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    626\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscores\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfloat32\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    629\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpvals_adj\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfloat64\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    630\u001b[0m }\n\u001b[1;32m    632\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m col \u001b[38;5;129;01min\u001b[39;00m test_obj\u001b[38;5;241m.\u001b[39mstats\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mlevels[\u001b[38;5;241m0\u001b[39m]:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'columns'"
     ]
    }
   ],
   "source": [
    "# We don't have an embedding for all genes, so we need to filter out these knockouts where we dont have the embedding:\n",
    "adata=ppp.map_and_filter_based_on_embedding(adata)\n",
    "\n",
    "# This calls DEGs and filters out treatments with to little cells\n",
    "adata=ppp.postprocess_anndata(adata, n_top_genes=2000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "722333d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# It writes/reads faster when not using compression\n",
    "adata.write(f'{save_path}/L1000_pp.h5ad', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9158b101",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f30c60d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed7d200a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a969ad97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "22188f58",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "load_path='/home/manu/chemCPA/chemCPA/anndatas'\n",
    "adata=sc.read(f'{load_path}/Norman_pp.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f4817f78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>treatment</th>\n",
       "      <th>treatment_dose_uM</th>\n",
       "      <th>treatment_time_h</th>\n",
       "      <th>treatment_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>control</th>\n",
       "      <th>pert_category</th>\n",
       "      <th>split</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGAAGC-1</th>\n",
       "      <td>K562</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>1</td>\n",
       "      <td>K562_control</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCATGTG-1</th>\n",
       "      <td>K562</td>\n",
       "      <td>TSC22D1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_TSC22D1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCCCTTG-1</th>\n",
       "      <td>K562</td>\n",
       "      <td>EKLF+MAP2K6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_EKLF+MAP2K6</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACGAAGCA-1</th>\n",
       "      <td>K562</td>\n",
       "      <td>control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>1</td>\n",
       "      <td>K562_control</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGACGTAG-1</th>\n",
       "      <td>K562</td>\n",
       "      <td>CEBPE+RUNX1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_CEBPE+RUNX1</td>\n",
       "      <td>ood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAGTACGT-8</th>\n",
       "      <td>K562</td>\n",
       "      <td>FOXA3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_FOXA3</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCACTCCA-8</th>\n",
       "      <td>K562</td>\n",
       "      <td>CELF2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_CELF2</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCCAACGG-8</th>\n",
       "      <td>K562</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_BCORL1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCTCCTAG-8</th>\n",
       "      <td>K562</td>\n",
       "      <td>PTPN12+ZBTB10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_PTPN12+ZBTB10</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGGCGAC-8</th>\n",
       "      <td>K562</td>\n",
       "      <td>MAP4K3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168.0</td>\n",
       "      <td>gene_knockout</td>\n",
       "      <td>Norman</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_MAP4K3</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111255 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   cell_line      treatment  treatment_dose_uM  \\\n",
       "index                                                            \n",
       "AAACCTGAGAAGAAGC-1      K562        control                NaN   \n",
       "AAACCTGAGGCATGTG-1      K562        TSC22D1                NaN   \n",
       "AAACCTGAGGCCCTTG-1      K562    EKLF+MAP2K6                NaN   \n",
       "AAACCTGCACGAAGCA-1      K562        control                NaN   \n",
       "AAACCTGCAGACGTAG-1      K562    CEBPE+RUNX1                NaN   \n",
       "...                      ...            ...                ...   \n",
       "TTTGTCATCAGTACGT-8      K562          FOXA3                NaN   \n",
       "TTTGTCATCCACTCCA-8      K562          CELF2                NaN   \n",
       "TTTGTCATCCCAACGG-8      K562         BCORL1                NaN   \n",
       "TTTGTCATCCTCCTAG-8      K562  PTPN12+ZBTB10                NaN   \n",
       "TTTGTCATCTGGCGAC-8      K562         MAP4K3                NaN   \n",
       "\n",
       "                    treatment_time_h treatment_type dataset  control  \\\n",
       "index                                                                  \n",
       "AAACCTGAGAAGAAGC-1             168.0  gene_knockout  Norman        1   \n",
       "AAACCTGAGGCATGTG-1             168.0  gene_knockout  Norman        0   \n",
       "AAACCTGAGGCCCTTG-1             168.0  gene_knockout  Norman        0   \n",
       "AAACCTGCACGAAGCA-1             168.0  gene_knockout  Norman        1   \n",
       "AAACCTGCAGACGTAG-1             168.0  gene_knockout  Norman        0   \n",
       "...                              ...            ...     ...      ...   \n",
       "TTTGTCATCAGTACGT-8             168.0  gene_knockout  Norman        0   \n",
       "TTTGTCATCCACTCCA-8             168.0  gene_knockout  Norman        0   \n",
       "TTTGTCATCCCAACGG-8             168.0  gene_knockout  Norman        0   \n",
       "TTTGTCATCCTCCTAG-8             168.0  gene_knockout  Norman        0   \n",
       "TTTGTCATCTGGCGAC-8             168.0  gene_knockout  Norman        0   \n",
       "\n",
       "                         pert_category  split  \n",
       "index                                          \n",
       "AAACCTGAGAAGAAGC-1        K562_control  train  \n",
       "AAACCTGAGGCATGTG-1        K562_TSC22D1  train  \n",
       "AAACCTGAGGCCCTTG-1    K562_EKLF+MAP2K6  train  \n",
       "AAACCTGCACGAAGCA-1        K562_control  train  \n",
       "AAACCTGCAGACGTAG-1    K562_CEBPE+RUNX1    ood  \n",
       "...                                ...    ...  \n",
       "TTTGTCATCAGTACGT-8          K562_FOXA3  train  \n",
       "TTTGTCATCCACTCCA-8          K562_CELF2  train  \n",
       "TTTGTCATCCCAACGG-8         K562_BCORL1  train  \n",
       "TTTGTCATCCTCCTAG-8  K562_PTPN12+ZBTB10  train  \n",
       "TTTGTCATCTGGCGAC-8         K562_MAP4K3  train  \n",
       "\n",
       "[111255 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "384f67bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52fd78dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fb9cc68",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "714fa784",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "g",
   "language": "python",
   "name": "g"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
